libcxx/include/__expected/expected.h

```diff
@@ -88,8 +88,289 @@ _LIBCPP_HIDE_FROM_ABI void __throw_bad_expected_access(_Arg&& __arg) {
 #  endif
 }
 
+struct __conditional_no_unique_address_invoke_tag {};
+
+// This class implements an object with `[[no_unique_address]]` conditionally applied to it,
+// based on the value of `_NoUnique`.
+//
+// A member of this class must always have `[[no_unique_address]]` applied to
+// it. Otherwise, the `[[no_unique_address]]` in the "`_NoUnique == true`" case
+// would not have any effect. In the `false` case, the `__v` is not
+// `[[no_unique_address]]`, so nullifies the effects of the "outer"
+// `[[no_unique_address]]` regarding data layout.
+//
+// If we had a language feature, this class would basically be replaced by `[[no_unique_address(condition)]]`.
+template <bool _NoUnique, class _Tp>
+struct __conditional_no_unique_address;
+
+template <class _Tp>
+struct __conditional_no_unique_address<true, _Tp> {
+  template <class... _Args>
+  _LIBCPP_HIDE_FROM_ABI constexpr explicit __conditional_no_unique_address(_Args&&... __args)
+      : __v(std::forward<_Args>(__args)...) {}
+
+  template <class _Func, class... _Args>
+  _LIBCPP_HIDE_FROM_ABI constexpr explicit __conditional_no_unique_address(
+      __conditional_no_unique_address_invoke_tag, _Func&& __f, _Args&&... __args)
+      : __v(std::invoke(std::forward<_Func>(__f), std::forward<_Args>(__args)...)) {}
+
+  _LIBCPP_NO_UNIQUE_ADDRESS _Tp __v;
+};
+
+template <class _Tp>
+struct __conditional_no_unique_address<false, _Tp> {
+  template <class... _Args>
+  _LIBCPP_HIDE_FROM_ABI constexpr explicit __conditional_no_unique_address(_Args&&... __args)
+      : __v(std::forward<_Args>(__args)...) {}
+
+  template <class _Func, class... _Args>
+  _LIBCPP_HIDE_FROM_ABI constexpr explicit __conditional_no_unique_address(
+      __conditional_no_unique_address_invoke_tag, _Func&& __f, _Args&&... __args)
+      : __v(std::invoke(std::forward<_Func>(__f), std::forward<_Args>(__args)...)) {}
+
+  _Tp __v;
+};
+
+// This function returns whether the type `_Second` can be stuffed into the tail padding
+// of the `_First` type if both of them are given `[[no_unique_address]]`.
+template <class _First, class _Second>
+inline constexpr bool __fits_in_tail_padding = []() {
+  struct __x {
+    _LIBCPP_NO_UNIQUE_ADDRESS _First __first;
+    _LIBCPP_NO_UNIQUE_ADDRESS _Second __second;
+  };
+  return sizeof(__x) == sizeof(_First);
+}();
+
 template <class _Tp, class _Err>
-class expected {
+class __expected_base {
```

Comment:
I think we should add a comment that explains the implementation of this class in pseudo-code, since it's kinda complex. We can say something like this:

```
This class implements the storage used by std::expected. We have a few goals for this storage:
1. Whenever the underlying {_Tp | _Unex} combination has free bytes in its tail padding, we should
   reuse it to store the bool discriminator of the expected, so as to save space.
2. Whenever the expected<_Tp, _Unex> as a whole has free bytes in its tail padding, we should
   allow an object following the expected to be stored in its tail padding.
3. However, we never want a user object (say `X`) that would follow an expected<_Tp, _Unex> to be stored in
   the padding bytes of the underlying {_Tp | _Unex} union, if any. That is because we use construct_at
   on that union, which would end up overwriting the `X` member if it is stored in the tail padding of the union.

To achieve this, [TODO]
```

Reply:
Good idea! To fill the "TODO" I could sum up some comments like https://github.com/llvm/llvm-project/pull/69673#issuecomment-1778553357 maybe.

Reply:
OK, I added a large comment to the `expected_base` class, taking your suggestion and summarizing some comments from the discussion. I also added some pseudo code of the `expected_base` class where the layout can be seen more easily.